crates/drag/src/platform_impl/gtk/mod.rs

```diff
@@ -97,25 +97,49 @@ fn clear_signal_handlers(window: &gtk::ApplicationWindow, handler_ids: &mut Vec<
     }
 }
 
-fn on_drop_cancel<F: Fn(DragResult) + Send + 'static>(
+fn on_drop_failed<F: Fn(DragResult, CursorPosition) + Send + 'static>(
     callback: Rc<F>,
     window: &gtk::ApplicationWindow,
     handler_ids: &Arc<Mutex<Vec<SignalHandlerId>>>,
-    drag_context: &gdk::DragContext,
 ) {
-    let window = window.clone();
-    let handler_ids = handler_ids.clone();
-
-    drag_context.connect_cancel(move |_, _| {
-        let handler_ids = &mut handler_ids.lock().unwrap();
-        clear_signal_handlers(&window, handler_ids);
-        window.drag_source_unset();
+    let window_clone = window.clone();
+    let handler_ids_clone = handler_ids.clone();
+
+    handler_ids
+        .lock()
+        .unwrap()
+        .push(window.connect_drag_failed(move |_, _, drag_result| {
+            let inhibit = match drag_result {
+                gtk::DragResult::NoTarget => {
+                    callback(
+                        DragResult::NoTarget,
+                        get_cursor_position(&window_clone).unwrap(),
+                    );
+                    Inhibit(true)
```

Comment:
I think this should be based on the callback return value (i.e. return true to prevent the default behavior)

Reply:
Hmm.. It's kind of awkward on mac. The animation shows before we call the callback function. Maybe we should make this an argument to the `start_drag` interface?